File: hydra-python/src/main/python/hydra/dsl/lib/eithers.py
"""Phantom-typed term DSL for the hydra.lib.eithers library."""

from hydra.phantoms import TTerm
from hydra.dsl.phantoms import lib_primitive as primitive, lib_primitive1 as primitive1, lib_primitive2 as primitive2, lib_primitive3 as primitive3


def either(f: TTerm, g: TTerm, e: TTerm) -> TTerm:
    """Eliminate an Either value by applying one of two functions."""
    return primitive3(f, g, e)


def is_left(e: TTerm) -> TTerm:
    """Check if an Either is a Left value."""
    return primitive1(e)


def is_right(e: TTerm) -> TTerm:
    """Check if an Either is a Right value."""
    return primitive1(e)


def map_left(f: TTerm, e: TTerm) -> TTerm:
    """Map a function over the Left side of an Either."""
    return primitive2(f, e)


def map_right(f: TTerm, e: TTerm) -> TTerm:
    """Map a function over the Right side of an Either."""
    return primitive2(f, e)


def from_left(default: TTerm, e: TTerm) -> TTerm:
    """Extract the Left value, or return a default."""
    return primitive2(default, e)


def from_right(default: TTerm, e: TTerm) -> TTerm:
    """Extract the Right value, or return a default."""
    return primitive2(default, e)

File: packages/hydra-typescript/src/main/typescript/hydra/lib/eithers.ts
/**
 * Either primitive functions for Hydra-TypeScript.
 */

import type { Either } from "../core.js";
import { left, right } from "../core.js";

export function fromLeft<L, R>(def: L, e: Either<L, R>): L {
  return e.tag === "left" ? e.value : def;
}

export function fromRight<L, R>(def: R, e: Either<L, R>): R {
  return e.tag === "right" ? e.value : def;
}

export function isLeft<L, R>(e: Either<L, R>): boolean {
  return e.tag === "left";
}

export function isRight<L, R>(e: Either<L, R>): boolean {
  return e.tag === "right";
}

export function mapLeft<L, L2, R>(
  f: (l: L) => L2,
  e: Either<L, R>,
): Either<L2, R> {
  return e.tag === "left" ? left(f(e.value)) : e;
}

export function mapRight<L, R, R2>(
  f: (r: R) => R2,
  e: Either<L, R>,
): Either<L, R2> {
  return e.tag === "right" ? right(f(e.value)) : e;
}
